pkg/sentry/fsimpl/iouringfs/iouringfs.go

```diff
@@ -0,0 +1,323 @@
+func (fd *fileDescription) SetStat(ctx context.Context, opts vfs.SetStatOptions) error {
+	return syserror.EFAULT
+}
+
+// Release implements vfs.FileDescriptionImpl.Release
+func (f *fileDescription) Release(ctx context.Context) {
+	// noop
+}
+
+// ConfigureMMap implements vfs.FileDescriptionImpl.ConfigureMMap.
+func (fd *fileDescription) ConfigureMMap(ctx context.Context, opts *memmap.MMapOpts) error {
+	opts.SentryOwnedContent = true
+	return vfs.GenericConfigureMMap(&fd.vfsfd, fd, opts)
+}
+
+// AddMapping implements memmap.Mappable.AddMapping.
+func (fd *fileDescription) AddMapping(ctx context.Context, ms memmap.MappingSpace, ar hostarch.AddrRange, offset uint64, writable bool) error {
+	fd.mappings.AddMapping(ms, ar, offset, writable)
+
+	return nil
+}
+
+// RemoveMapping implements memmap.Mappable.RemoveMapping.
+func (fd *fileDescription) RemoveMapping(ctx context.Context, ms memmap.MappingSpace, ar hostarch.AddrRange, offset uint64, writable bool) {
+	fd.mappings.RemoveMapping(ms, ar, offset, writable)
+}
+
+// CopyMapping implements memmap.Mappable.CopyMapping.
+func (fd *fileDescription) CopyMapping(ctx context.Context, ms memmap.MappingSpace, srcAR, dstAR hostarch.AddrRange, offset uint64, writable bool) error {
+	return fd.AddMapping(ctx, ms, dstAR, offset, writable)
+}
+
+// InvalidateUnsavable implements memmap.Mappable.InvalidateUnsavable.
+func (*fileDescription) InvalidateUnsavable(context.Context) error {
+	panic("InvalidateUnsavable not implemented")
+	return nil
+}
+
+func (fd *fileDescription) sqEndOffset() uint64 {
+	// Sizeof need an instance of the type...
+	var forSizeof uint
+	SQEndOff := uint64(fd.sqEntries) * uint64(unsafe.Sizeof(forSizeof))
+	return fs.OffsetPageEnd(int64(linux.IORING_OFF_SQ_RING + SQEndOff))
+}
+
+func (fd *fileDescription) cqEndOffset() uint64 {
+	// Sizeof need an instance of the type...
+	cqe := linux.IoUringCqe{}
+	CQEndOff := uint64(fd.cqEntries) * uint64(unsafe.Sizeof(cqe))
+	return fs.OffsetPageEnd(int64(linux.IORING_OFF_CQ_RING + CQEndOff))
+}
+
+func (fd *fileDescription) sqeEndOffset() uint64 {
+	// Sizeof need an instance of the type...
+	sqe := linux.IoUringSqe{}
+	SQESEndOff := uint64(fd.sqEntries) * uint64(unsafe.Sizeof(sqe))
+	return fs.OffsetPageEnd(int64(linux.IORING_OFF_SQES + SQESEndOff))
+}
+
+func (fd *fileDescription) isInRing(maprange memmap.MappableRange) bool {
+	if maprange.Start >= linux.IORING_OFF_SQ_RING && maprange.End <= fd.sqEndOffset() {
+		return true
+	}
+
+	if maprange.Start >= linux.IORING_OFF_CQ_RING && maprange.End <= fd.cqEndOffset() {
+		return true
+	}
+
+	if maprange.Start >= linux.IORING_OFF_SQES && maprange.End <= fd.sqeEndOffset() {
+		return true
+	}
+
+	return false
+
+}
+
+// Translate implements memmap.Mappable.Translate.
+func (fd *fileDescription) Translate(ctx context.Context, required, optional memmap.MappableRange, at hostarch.AccessType) ([]memmap.Translation, error) {
+	// Sq ring, Cq ring or Sqe ring ?
+	if fd.isInRing(required) {
```

Comment:
Invert the check and return early if `required` is not in the ring. Removes one level of branching from the body the function.

```
if !fd.isInRing(required) {
    return nil, &memmap.BusError{io.EOF}
}

// Allocation, create translations, etc...
```

Reply:
Yes, good idea. Fixed